Fix decay restarts

rgknox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description:

This is a bugfix. These changes are required to ensure b4b comparisons between a base and its restart.  This is due to a recent regression from a very recent tag.

### Expectation of Answer Changes:

This will change the answers following a restart, particularly those answers related to soil carbon, as this will modify the litter flux slightly.  

### Checklist:
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [x] FATES PASS/FAIL regression tests were run


### Test Results:
<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

TBD

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

ckoven reviewed (approved):
this looks good to me, thanks @rgknox for catching.

rgknox commented:
Tests on cheyenne:

/glade/scratch/rgknox/clmed-tests/fix-decay-ctsm5.1.dev041

One noted difference, the PRT2 test generated very small differences with the base, see here:

/glade/scratch/rgknox/ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCru.cheyenne_intel.clm-FatesPRT2.GC.20210618_075109_ll9h81/run/ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCru.cheyenne_intel.clm-FatesPRT2.GC.20210618_075109_ll9h81.clm2.h0.2000-01-01-00000.nc.cprnc.out

The largest rms difference was on the order of e-21.  This might be fine, but I'd like to think through this a little more.

rosiealice commented on biogeochem/FatesSoilBGCFluxMod.F90:
```diff
@@ -997,9 +997,9 @@ subroutine FluxIntoLitterPools(csite, bc_in, bc_out)
-          bc_out%litt_flux_lab_c_si(:) = 0._r8
+          bc_out%litt_flux_cel_c_si(:) = 0.0_r8
+          bc_out%litt_flux_lig_c_si(:) = 0.0_r8
+          bc_out%litt_flux_lab_c_si(:) = 0.0_r8
```
Embarrassing question I've never got my head around. Does this matter? Like, are there circumstances where you get a different answer with a 0.0_r8 than a 0._r8?

rgknox replied:
nope, I don't think it matters.  This was just an artifact too, I didn't make the change for stylistic reasons.  This is what happened.  I was testing the code by forcing a non zero decomposition flux into E3SM with a value of 1.0. I wanted to see if E3SM was closing its carbon budget with a constant and known input flux from FATES. When I reverted it, I just added a zero.

rosiealice commented on main/FatesRestartInterfaceMod.F90:
```diff
@@ -944,6 +946,17 @@ subroutine define_restart_vars(this, initialize_variables)
 
+
+    call this%RegisterCohortVector(symbol_base='fates_seed_decay', vtype=cohort_r8, &
+            long_name_base='seed bank (non-germinated)',  &
```
Is this 'long name' accurate? Like, if it's the seed decay rate, it's not actually in the seed bank any more?

rgknox replied:
good catch, that is not the right long name. Its a flux rate. I stole the long_name from the pool variable that was already there and forgot to update the name.  Will do.

rosiealice commented on main/FatesRestartInterfaceMod.F90:
```diff
@@ -944,6 +946,17 @@ subroutine define_restart_vars(this, initialize_variables)
+
+    call this%RegisterCohortVector(symbol_base='fates_seedgerm_decay', vtype=cohort_r8, &
+           long_name_base='seed bank (germinated)',  &
+           units='kg/m2', veclength=num_elements, flushval = flushzero, &
```
Relatedly, while I know this isn't really related to this bug fix, without delving into the code a bit more, I don't think it's super clear from the names how these variables differ.

rgknox replied:
will update this as well, thanks

rosiealice reviewed (approved):
Comments here entirely related to my own slight incomprehension of how this code has evolved. Otherwise it looks fine.

rgknox commented:
I reverted a portion of the code where the labile fluxes to the HLM were calculated. The diagnostic site level array with effluxes is populated after the restart when we set the fragmentation fluxes, so it should work. I reverted it to use the cohort level effluxes, to preserve b4b with the test which seems safer.  I think the difference could be explained by order of operations rounding errors (which are acceptable).

rgknox commented:
retesting

rgknox commented:
tests pass: /glade/scratch/rgknox/clmed-tests/fix-decay-ctsm5.1.dev041-v4
